doc/design/Notifications.md

```diff
@@ -58,9 +58,35 @@ An inattentive User could configure a device to cause a serious battery drain pr
 
 The service approach is necessarily more complex than the server approach but not overly so.  Much of the application's code can be shared.  And much of the service code is standard practice with Android.  The Firebase JobDispatcher library provides a compatible facility to support job scheduling.  In recent devices the JobDisatcher uses Android's JobSchedular class to efficiently manage the battery use.
 
+Both this approach and the server approach rely on polling via database change listeners to trigger notifications.
+
 ### Conclusion
 
-Eliminating the need for a notification server, creating one less moving part (and a big moving part at that) and the ability to reuse and localize code with the app itself is compelling, even with the risk of battery abuse and more complex code (especially User configuration and job scheduling).  Thus, the service approach carries the day, in our humble opinion.
+Eliminating the need for a notification server, creating one less moving part (and a big moving part at that) and the ability to reuse and localize code with the app itself is compelling, even with the risk of battery abuse and more complex code (especially User configuration and job scheduling).  Thus, the service approach is more appealing than the server approach.
+
+## Cloud Functions Approach
+
+The Google Firebase team has provided a light-weight capability to execute (JavaScript using NodeJS) functions when certain database events take place.  These functions are developed separately from the app and deployed to a Firebase project using a NodeJS framework/infrastructure supplied by the Firebase team.
+
+### Ramifications and Conclusion
+
+There is a cost to using the Firebase Cloud Functions (FCF) albeit small.  More importantly, FCF is another moving part separate from the application.  While it is simpler than both a hosted server and avoids all the problems with a background service, FCF is still not an optimal solution.
+
+## Direct Notifications Approach
+
+In this approach the app drives notifications directly.  On the posting of a new message or a game move (or partial move) the app sends push notifcations to the FCM servers using the HTTP method (as opposed to XMPP).
+
+### Overview
+
+To implement direct notifications, each room profile maintains a list of device tokens associated with the members of the room.  When a new message is posted or a game move occurs push notifications are queued up to be sent by a background service (MainService) to each room.  A device running the app in the foreground processes the notification to determine it should be added to the status bar, otherwise the Android framework diretly adds the notification to the status bar.
```

Comment:
Last sentence: "to determine *if* it should be added..", typo in "diretly". Also, the final thought is not clear: the app determines if the notification should by added to the status bar and otherwise adds it to the status bar.